[![CI](https://github.com/microbemax/dnabactanalyzer/actions/workflows/ci.yaml/badge.svg)](https://github.com/microbemax/dnabactanalyzer/actions/workflows/ci.yaml)
![License: MIT](https://img.shields.io/badge/license-MIT-green)
![Python 3.9](https://img.shields.io/badge/python-3.9-blue.svg)

# GenBankProcessor

`GenBankProcessor` is a Python class designed to process GenBank files and extract useful genomic information. It allows users to parse GenBank files, extract various genomic features, and write sequences to files in different formats.

## Features

- Parse GenBank files to extract genomic features.
- Write full nucleotide sequences to FASTA format.
- Compile extracted genomic features into a DataFrame for easy manipulation and analysis.


# NucleotideContent

`NucleotideContent` is a Python class designed for analyzing and visualizing nucleotide content in DNA sequences. It reads DNA sequences from a FASTA file, calculates GC and AT content, and generates WIG files for visualization. This class is particularly useful for handling large sequences and applying smoothing to analyze nucleotide distribution more clearly.

## Features

- Read DNA sequences from FASTA files.
- Calculate GC and AT content.
- Generate WIG files for visualization of nucleotide content.
- Apply smoothing for clearer analysis of nucleotide distribution.


## Installation

These modules not not require a separate installation if you have the prerequisites installed. Ensure you have Python 3.9 or higher. The requirements.txt file is provided so you can install it with pip

